### PR TITLE
Add quotes

### PR DESCRIPTION
**Pull request by Murph33**
Very basic page.  Anyone can add a quote.  No CSS.  No users.  All the tests are green though!

### CODE REVIEW

File: db/seeds.rb [hunk 1]
```diff
@@ -5,3 +5,12 @@
 #
 #   cities = City.create([{ name: 'Chicago' }, { name: 'Copenhagen' }])
 #   Mayor.create(name: 'Emanuel', city: cities.first)
+100.times do 
+
+  Quote.create author: Faker::Name.first_name,
```
Comment (jhawthorn) [inline db/seeds.rb]: Indentation and spacing is a little awkward in this file. `create!` would also be better than `create` here. `create` can fail to save (if validations fail) and you would have to check the returned object to know. Instead there is `create!` which will raise an exception if it wasn't successful.

I'd rather see

``` ruby
100.times do
  Quote.create!(
    author: Faker::Name.first_name,
    body:  Faker::Lorem.sentences.join(" "),
    location: Faker::Address.street_address
  )
end
```

File: app/views/quotes/show.html.erb [hunk 2]
```diff
@@ -0,0 +1 @@
+<%= @quote.author %> said, "<%= @quote.body %>", in the <%= @quote.location %>
```
Comment (jhawthorn) [inline app/views/quotes/show.html.erb]: This is the same as the line in `app/views/quotes/index.html.erb` above. We can share the formatting here by using a partial http://guides.rubyonrails.org/layouts_and_rendering.html#using-partials. This could allow us in the future to make changes in one place and have them apply to both pages. We'll likely want to do this when we're adding styles.

Having a partial like `app/views/quotes/_quote.html.erb` and then rendering that from both `show.html.erb` and `index.html.erb`. This pattern and naming is so common that rails supports a shortcut for it.

``` erb
<%= render @quote %>
```

which does the same thing as

``` erb
<%= render 'quote', quote: @quote %>
```

**Comment by adammathys** [comment]
In an ideal world, we would break this pull-request into multiple smaller ones. Small commits are great, but so are reasonably sized pull-requests!

By my count, this could easily be 4 pull-requests:
1. Add and configure RSpec.
2. Add RuboCop config.
3. Add simple_form and config.
4. Add new model and controller for quotes.

If we organized our pull-requests that way, we could easily merge in the smaller chunks without having to hold them back if any of the others require further changes.

**Comment by adammathys** [comment]
Several of the commits contain more work than would ideally be included in a single commit. 
- [Add Rspec. Add quote model](https://github.com/StemboltHQ/quoteboard/commit/aca68451e01255f4ddee0f004c8771506c117d4a)
  Should be broken into at least two commits.
- [Add quote model test](https://github.com/StemboltHQ/quoteboard/commit/0f5e1d9c6d1625083ed875c3bc97d7d9430b4bce)
  Shouldn't exists. Most of the work should be part of the commit that adds the model. With a separate commit to add Faker. 
- [Add tests for quote new action](https://github.com/StemboltHQ/quoteboard/commit/dd87c2e38c2e2f4b6a6f978d6220d029c32e3cff)
  Moving gems should have been it's own commit. No need for empty `.scss`, `.coffee`, and `_helper.rb` Adds a test for a non-existent action. Every commit should ideally have passing tests.
- [Add more quotes tests](https://github.com/StemboltHQ/quoteboard/commit/72452480fb08b12c837f7d6b4c4dddbefd5f2e88)
  Adds more tests for undefined actions. Has remnants of a merge conflict. Contains changes to `spec_helper` as well.
- [Add more quotes tests](https://github.com/StemboltHQ/quoteboard/commit/f6e0e6f0d07065a9d185a19b8e581f5198972c75)
  Doesn't actually add tests. It adds the undefined actions we wrote tests for in the previous two commits. Adding those actions should have happened in the same commits we added tests for them. Cleans up our merge conflict remnants, which wouldn't need to happen if we fixed it in the original commit.
- [Add quote actions and routes](https://github.com/StemboltHQ/quoteboard/commit/ad59b97b7e583c0787eb16355bd42c09981ec6ef)
  Only adds routes. I could be mistaken, but I don't think any of our tests will pass without the routes. We should have been adding the routes along with the actions.
- [Fix quote actions, pass tests](https://github.com/StemboltHQ/quoteboard/commit/e352426a625bf80758beda75c3f6e708af2a8b88)
  Finally fixes everything so that tests pass. All of these changes should be part of previous commits and not a generic "Fix everything" commit afterwards.
- [Set root. Fix seeds. Show index](https://github.com/StemboltHQ/quoteboard/commit/432a2e77ff2875865993a97025efa6cf4eb071dd)
  Three separate tasks here. Should be three separate commits. 
- [Add simple_form. Add new quote form](https://github.com/StemboltHQ/quoteboard/commit/651df9c59c2220301e8b343c96c1da284e379f66)
  Two separate tasks. Split into separate commits.
- [Add layout links. Update views.](https://github.com/StemboltHQ/quoteboard/commit/2f67f8421101556dfcc797b3c71def7846760115)
  Has a fix for a previous commit. Then edits three different actions.
- [Add a single space in show page](https://github.com/StemboltHQ/quoteboard/commit/c9b7a6200037df9c1c754b423487d477efdd1920)
  Fixes previous commit. Doesn't need to be it's own commit.
- [Modify index and show views.](https://github.com/StemboltHQ/quoteboard/commit/bf4ba822169fde63edf48a4beb8d4da88bce33f8)
  More fixes to previous commits. Should just be squashed into the original commits.
- [Add rubocop configuration. Add config settings.](https://github.com/StemboltHQ/quoteboard/commit/31c6b5adc9b21e9ef66de49fb822c2060df65c8f)
  Adds RuboCop (:+1:) and a duplicate configuration entry (:-1:).

Reading through this list, I would ideally like to see ~10 commits:
1. Add RSpec & config.
2. Add Faker.
3. Changes to Gemfile & Gemfile.lock
4. Add RuboCop config.
5. Create `Quote` model with factory and tests. 
6. Add simple_form & config.
7. Add `QuotesController#new` & `QuotesController#create` actions, views and tests.
8. Add `QuotesController#show` & `QuotesController#index` actions, views and tests.
9. Add `QuotesController#edit` & `QuotesController#update` actions, views and tests.
10. Add `QuotesController#destroy` action and tests.

The last 4 commits could be combined into a single one or split up further.

Side note: All of these commits contain a `Change-Id:`. Which is a Gerrit specific thing that should be stripped out. They're automatically added by `.git/hooks/commit-msg`. So delete that if you haven't already and then you can remove them from these commits without it getting added back.

File: app/controllers/quotes_controller.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
+class QuotesController < ApplicationController
+  def show
+    @quote = Quote.find params[:id]
```
Comment (Dkendal) [inline app/controllers/quotes_controller.rb:3]: What behaviour would you expect here if you entered a url with an `id` that did not match any model in the database?

File: app/views/layouts/application.html.erb [hunk 4]
```diff
@@ -7,7 +7,10 @@
   <%= csrf_meta_tags %>
 </head>
 <body>
-
+  <div>
+    <%= link_to "Add Quote", new_quote_path %>
+    <%= link_to "All Quotes", quotes_path %>
+  </div>
 <%= yield %>
```
Comment (Dkendal) [inline app/views/layouts/application.html.erb]: Inconsistent indentation: should be 2 spaces.

File: app/views/quotes/_quote.html.erb [hunk 5]
```diff
@@ -0,0 +1 @@
+<%= quote.author %> said, "<%= quote.body %>", in the <%= quote.location %>
```
Comment (Dkendal) [inline app/views/quotes/_quote.html.erb]: Suggestion for the future: html has has `<q></q>` tag for inline quotes, as well as `<blockquote></blockquote>` for (as the name implies) block quotes.

File: spec/controllers/quotes_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+      it "changes the quotes count by +1" do
+        expect { subject }.to change { Quote.count }.by(1)
+      end
+
+      it "redirects to the show page" do
+        expect(subject).to redirect_to quote_path(Quote.last)
+      end
+    end
+    context "with invalid parameters" do
+      subject { post :create, quote: attributes_for(:quote).merge(body: nil) }
```
Comment (Dkendal) [inline spec/controllers/quotes_controller_spec.rb]: `quote: attributes_for(:quote).merge(body: nil)`
is the same as
`quote: attributes_for(:quote, body: nil)`

File: spec/controllers/quotes_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+    it "instantiates a new Quote variable" do
+      subject
+      expect(assigns(:quote)).to be_a_new Quote
+    end
+  end
+
+  describe "#create" do
+    context "with valid parameters" do
+      subject { post :create, quote: attributes_for(:quote) }
```
Comment (Dkendal) [inline spec/controllers/quotes_controller_spec.rb]: I would probably pull the params out here.

``` ruby
subject { post :create, params }
let(:params) { quote: attributes_for(:quote) }

...

context "with invalid parameters" do
  let(:params) { quote: {} }
  ...
end
...
```

File: spec/controllers/quotes_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        expect(subject).to render_template(:new)
+      end
+
+      it "doesn't change the quotes count" do
+        expect { subject }.to_not change { Quote.count }
+      end
+    end
+  end
+
+  describe "#edit" do
```
Comment (Dkendal) [inline spec/controllers/quotes_controller_spec.rb]: Empty test.

File: spec/controllers/quotes_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+  end
+
+  describe "#index" do
+    subject { get :index }
+    it "renders the index template" do
+      expect(subject).to render_template(:index)
+    end
+
+    context "more than one quote in database" do
+      let!(:quote_1) { create(:quote) }
```
Comment (Dkendal) [inline spec/controllers/quotes_controller_spec.rb]: A more meaningful variable name than `quote_1` would be better here. People aren't very good at keeping numbers in their heads but we can manage nouns.

File: spec/controllers/quotes_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+      before { quote }
+      it "instantiates a quotes variable containing all quotes" do
+        subject
+        expect(assigns(:quotes)).to eq([quote_1, quote])
```
Comment (Dkendal) [inline spec/controllers/quotes_controller_spec.rb]: I don't think creating an additional quote is necessary here, the behaviour isn't any different then with a single record.
can just be 

`expect(assigns(:quotes)).to eq([quote])`

then you can get rid of the context.

In general the usage of `context` is implying that there is some kind of branching behaviour, if I don't see a context, and then another that is complementary I'm going to be very confused.

File: spec/factories/quotes.rb [hunk 11]
```diff
@@ -0,0 +1,7 @@
+FactoryGirl.define do
+  factory :quote do
+    author   Faker::Name.first_name
```
Comment (Dkendal) [inline spec/factories/quotes.rb]: You'll want to make these attributes lazy evaluated by using a block, or else every quote will have the name author , body and location. This likely doesn't matter if there is no uniqueness constraint but it may be confusing or unintended. 

e.g.
`author { Faker::Name.first_name }`

File: app/views/quotes/index.html.erb [hunk 12]
```diff
@@ -0,0 +1,6 @@
+<% @quotes.each do |quote| %>
+  <div>
```
Comment (graygilmore) [inline app/views/quotes/index.html.erb]: Similar to @Dkendal's note here: https://github.com/StemboltHQ/quoteboard/pull/1/files#r48580873, instead of a `<div>` you could use a `<blockquote>`.

Comment (graygilmore) [inline app/views/quotes/index.html.erb]: Upon further thinking, you're probably better off including `<blockquote>` in the partial itself.

File: spec/factories/quotes.rb [hunk 13]
```diff
@@ -0,0 +1,7 @@
...
+    body     Faker::Lorem.sentences.join(" ")
```
Comment (Dkendal) [inline spec/factories/quotes.rb]: We tend to avoid column aligning text like this as it ends up creating more diffs in commits than were necessary.

File: db/seeds.rb [hunk 14]
```diff
@@ -5,3 +5,10 @@
 #
 #   cities = City.create([{ name: 'Chicago' }, { name: 'Copenhagen' }])
 #   Mayor.create(name: 'Emanuel', city: cities.first)
+100.times do
+  Quote.create!(
+    author: Faker::Name.first_name,
+    body:  Faker::Lorem.sentences.join(" "),
```
Comment (graygilmore) [inline db/seeds.rb]: There's an extra space after `body:`

File: app/controllers/quotes_controller.rb (hunk 3 above)
Comment (Murph33) [inline app/controllers/quotes_controller.rb:3]: Throw an exception.  I guess better UX would be using find_by and then checking if we have a person object or nil `unless @quote.nil?` or something and redirecting to either the quote or a no quote page depending on if it's there or not.

EDIT

rubocop is not happy with unless/else.  Fair enough, life.

Comment (Dkendal) [inline app/controllers/quotes_controller.rb:3]: I would also expect it to throw an error. `find` will return `nil` if no record is found, whereas `find!` will throw `ActiveRecord::RecordNotFound`. Rails will rescue this exception with a 404 page by default.

Comment (Murph33) [inline app/controllers/quotes_controller.rb:3]: `find` throws an error if no record is found.  `find_by` returns nil if no record is found.  Unless I'm misunderstanding something.

File: app/views/quotes/index.html.erb [hunk 15]
```diff
@@ -0,0 +1,4 @@
+<% @quotes.each do |quote| %>
+    <%= render quote %>
```
Comment (graygilmore) [inline app/views/quotes/index.html.erb]: Indented too far here.

File: app/views/quotes/_quote.html.erb [hunk 16]
```diff
@@ -0,0 +1 @@
+<div><%= quote.author %> said, <q><%= quote.body %></q>, in the <%= quote.location %></div>
```
Comment (graygilmore) [inline app/views/quotes/_quote.html.erb:1]: No desire to use `blockquote`?

File: config/routes.rb [hunk 17]
```diff
@@ -1,3 +1,5 @@
 Rails.application.routes.draw do
-
+  root "quotes#index"
+  resources :quotes
+  get '/quotes/error' => 'quotes#error'
```
Comment (graygilmore) [inline config/routes.rb]: I believe the preferred format is now:

``` ruby
get '/quotes/error', to: 'quotes#error'
```

File: app/views/quotes/_quote.html.erb (hunk 16 above)
Comment (Murph33) [inline app/views/quotes/_quote.html.erb:1]: My logic was that we likely won't have very long quotes being uploaded to here.  Could add something to the model so that could toggle between the two depending on user preference.

Comment (graygilmore) [inline app/views/quotes/_quote.html.erb:1]: Ah! I see what you're doing. I missed the `<q>`. :+1:

File: .gitignore [hunk 18]
```diff
@@ -8,6 +8,7 @@
 /.bundle
 /Gemfile.lock
 /*.un~
+/.bye*
```
Comment (cbrunsdon) [inline .gitignore]: Doesn't belong in this commit.

Also generally would belong in a global git ignore, not in this file. As with most of the things in this file.

File: app/controllers/quotes_controller.rb [hunk 19]
```diff
@@ -0,0 +1,36 @@
+class QuotesController < ApplicationController
+  def show
+    @quote = Quote.find_by(id: params[:id])
+    if @quote.nil?
+      render :error
```
Comment (cbrunsdon) [inline app/controllers/quotes_controller.rb]: this is probably more functionality than you need for now, to be honest. Just throwing the exception or dealing with some kind of flash notice later would have been fine for a commit #1

Comment (Dkendal) [inline app/controllers/quotes_controller.rb]: to clarify @Murph33 I was suggesting using `find!` and letting rails render the 404 automatically

File: .gitignore (hunk 18 above)
Comment (Murph33) [inline .gitignore]: Fixed up my global ignore and rewrote history removing Gemfile.lock (I have no idea why I added this ever) and removing *.un~ as well.

File: app/controllers/quotes_controller.rb (hunk 19 above)
Comment (Murph33) [inline app/controllers/quotes_controller.rb]: I think `find!` is deprecated now?

```
Quote.respond_to? :find!
=> false
```

Maybe I'm not seeing something but the difference in the error page rails shows between using `find`, `find_by`, and `find_by!` seems to be very slight

Comment (Dkendal) [inline app/controllers/quotes_controller.rb]: Sorry, I was mistaken - `find` will raise an error if the record isn't found. just using `find` is sufficient
